pcp/PCPGenericData.c

```diff
@@ -0,0 +1,321 @@
+   Process_printCount(str, val, coloring);
+
+static void PCPGenericData_writeField(const GenericData* this, RichString* str, int field) {
+   const PCPGenericData* gg = (const PCPGenericData*) this;
+   PCPGenericDataField* gf = (PCPGenericDataField*)Hashtable_get(gg->fields, field);
+   if (!gf)
+      return;
+
+   const ProcessField* fields = this->settings->ss->fields;
+   char buffer[256];
+   int attr = CRT_colors[DEFAULT_COLOR];
+   bool coloring = this->settings->highlightMegabytes;
+
+   DynamicColumn* dc = Hashtable_get(this->settings->dynamicColumns, fields[field]);
+   if (!dc || !dc->enabled)
+      return;
+
+   PCPDynamicColumn* column = (PCPDynamicColumn*) dc;
+   bool instances = column->instances;
+
+   bool scaled = column->scale;
+   pmAtomValue atom;
+   pmUnits kbyte_scale;
+
+   kbyte_scale.dimSpace = 1;
+   kbyte_scale.scaleSpace = PM_SPACE_KBYTE;
+
+   pmUnits sec_scale;
+   sec_scale.dimTime = 1;
+   sec_scale.scaleTime = PM_TIME_SEC;
+
+   pmUnits count_scale;
+   count_scale.dimCount = 1;
+   count_scale.scaleCount = PM_COUNT_ONE;
+
+   int width = column->super.width;
+   if (!width || abs(width) > DYNAMIC_MAX_COLUMN_WIDTH)
+      width = DYNAMIC_DEFAULT_COLUMN_WIDTH;
+   int abswidth = abs(width);
+   if (abswidth > DYNAMIC_MAX_COLUMN_WIDTH) {
+      abswidth = DYNAMIC_MAX_COLUMN_WIDTH;
+      width = -abswidth;
+   }
+
+   if (instances) {
+      char* instName;
+      attr = CRT_colors[DYNAMIC_GRAY];
+
+      PCPMetric_externalName(gf->pmid, gf->interInst, &instName);
+
+      xSnprintf(buffer, sizeof(buffer), "%*.*s ", width, 250, instName);
+      RichString_appendAscii(str, attr, buffer);
+   } else {
+      switch (gf->type) {
+         case PM_TYPE_STRING:
+            attr = CRT_colors[DYNAMIC_GREEN];
+            xSnprintf(buffer, sizeof(buffer), "%*.*s ", width, 250, gf->value->cp);
+            RichString_appendAscii(str, attr, buffer);
+            break;
+         case PM_TYPE_32:
+            if (scaled) {
+               if (gf->units.dimSpace) {
+                  KBytesField(PM_TYPE_32, atom.l);
+                  break;
+               } else if (gf->units.dimTime) {
+                  TimeField(PM_TYPE_32, atom.l);
+                  break;
+               } else {
+                  CountField(PM_TYPE_32, atom.l);
+                  break;
+               }
+            } else {
+               PrintField("%*d ", gf->value->l);
+               break;
+            }
+         case PM_TYPE_U32:
+            if (scaled) {
+               if (gf->units.dimSpace) {
+                  KBytesField(PM_TYPE_U32, atom.ul);
+                  break;
+               } else if (gf->units.dimTime) {
+                  TimeField(PM_TYPE_U32, atom.ul);
+                  break;
+               } else {
+                  CountField(PM_TYPE_U32, atom.ul);
+                  break;
+               }
+            } else {
+               PrintField("%*u ", gf->value->ul);
+               break;
+            }
```

Comment:
This looks very repetetive in the following blocks … Only thing changing is the accessed field and the `PM_TYPE_*` specifier. Care to use a `define` for this?

Reply:
This was resolved with the refactoring awhile back that improved the generic value scaling algorithm used.  Code referred to here now lives in PCPDynamicColumn.c FWIW.